### PR TITLE
PMM-12712 Shards collector.

### PR DESCRIPTION
**Pull request by JiriCtvrtka**
[PMM-12712](https://jira.percona.com/browse/PMM-12712)

https://github.com/Percona-Lab/pmm-submodules/pull/3485

https://github.com/percona/pmm/pull/2701

[PMM-12712]: https://perconadev.atlassian.net/browse/PMM-12712?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by JiriCtvrtka** [comment]
I am trying to figure out why tests failing, but localy it is fine.

File: docker/scripts/init-shard.sh [hunk 1]
```diff
@@ -1,5 +1,14 @@
 #!/bin/bash 
 
+readarray -d . -t verarr <<< "${VERSION}"
+readarray -d : -t version <<< "${verarr[0]}"
+echo "Mongo version: ${version[1]}"
+
+if ((${version[1]} <= 4))
+then
+  echo -e 'mongo' > /usr/bin/mongosh && chmod +x /usr/bin/mongosh
```
Comment (BupycHuk) [inline docker/scripts/init-shard.sh:9]: Maybe it's better to use this approach https://github.com/supercharge/mongodb-github-action/blob/main/start-mongodb.sh#L11-L30

File: exporter/sharded_collector.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+)
+
+type shardedCollector struct {
```
Comment (BupycHuk) [inline exporter/sharded_collector.go:30]: `shardsCollector`?

File: docker/scripts/init-shard.sh (hunk 1 above)
Comment (JiriCtvrtka) [reply docker/scripts/init-shard.sh:9]: Yes this looks better, ok I will change it.

File: exporter/sharded_collector.go (hunk 2 above)
Comment (JiriCtvrtka) [reply exporter/sharded_collector.go:30]: Ok lets rename it. Should I apply it everywhere? Even on PMM PR?

Comment (BupycHuk) [reply exporter/sharded_collector.go:30]: Yes

Comment (JiriCtvrtka) [reply exporter/sharded_collector.go]: Done in both PRs.

File: exporter/shards_collector.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+	rs, err := cursor.Find(d.ctx, bson.M{"_id": bson.M{"$regex": fmt.Sprintf("^%s.", database), "$options": "i"}})
+	if err != nil {
+		logger.Errorf("cannot find _id starting with \"%s.\":%s", database, err)
+		return nil
+	}
+
+	var decoded []bson.M
+	err = rs.All(d.ctx, &decoded)
+	if err != nil {
+		logger.Errorf("cannot decode collections:%s", err)
```
Comment (idoqo) [inline exporter/shards_collector.go:144]:
```suggestion
		logger.Errorf("cannot decode collections: %s", err)
```

**Review by idoqo — APPROVED** [review]

**Review by BupycHuk — APPROVED** [review]

**Pull request merged**